### PR TITLE
Moving fipa-independant code to the base dialogue module

### PR DESCRIPTION
**Pull request by 5A11**
## Proposed changes

This PR moves all fipa-independant code from the fipa dialogue module to the base dialogue module.

## Types of changes

What types of changes does your code introduce to agents-aea?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist

_Put an `x` in the boxes that apply._

- [x] I have read the [CONTRIBUTING](../CONTRIBUTING.md) doc
- [ ] I am making a pull request against the `develop` branch (left side). Also you should start your branch off our `develop`.
- [ ] Lint and unit tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have checked that code coverage does not decrease.
- [ ] I have checked that the documentation about the `aea cli` tool works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: packages/fetchai/protocols/fipa/dialogues.py [hunk 1]
```diff
@@ -80,64 +113,69 @@ class AgentRole(Dialogue.Role):
...
-        :return: The role of the agent
+        :return: the agent's role
         """
         fipa_message = cast(FipaMessage, message)
-        return role_from_first_message(fipa_message)
+        if fipa_message.is_incoming:
+            return FipaDialogue.AgentRole.SELLER
+        else:
+            return FipaDialogue.AgentRole.BUYER
+        # return role_from_first_message(fipa_message)
```
Comment (5A11) [inline packages/fetchai/protocols/fipa/dialogues.py:127]: I had to put the code here for now due to circular import issue in custom_types.py

File: packages/fetchai/protocols/fipa/custom_types.py [hunk 2]
```diff
@@ -24,11 +24,32 @@
 from aea.helpers.search.models import Query as BaseQuery
 from aea.protocols.base import Message
 
+# from packages.fetchai.protocols.fipa.dialogues import FipaDialogue
+
```
Comment (5A11) [inline packages/fetchai/protocols/fipa/custom_types.py:28]: See circular import issue comment below

File: packages/fetchai/protocols/fipa/custom_types.py [hunk 3]
```diff
@@ -24,11 +24,32 @@
...
 def role_from_first_message(message: Message):
-    pass
+    """
+    Infer the role of the agent from an incoming or outgoing first message
+
+    :param message: an incoming/outgoing first message
+    :return: the agent's role
+    """
+    raise NotImplementedError
+    # return FipaDialogue.AgentRole.BUYER
```
Comment (5A11) [inline packages/fetchai/protocols/fipa/custom_types.py:43]: See circular import issue comment below

File: packages/fetchai/protocols/fipa/dialogues.py [hunk 4]
```diff
@@ -194,185 +232,12 @@ def __init__(self, agent_address: Address) -> None:
 
         :return: None
         """
-        Dialogues.__init__(self, agent_address=agent_address)
+        Dialogues.__init__(
+            self, concrete_dialogue_class=FipaDialogue, agent_address=agent_address
```
Comment (5A11) [inline packages/fetchai/protocols/fipa/dialogues.py:236]: Needed this for the base class `create_dialogue` methods.

File: aea/helpers/dialogue/base.py [hunk 5]
```diff
@@ -378,22 +538,47 @@ def __str__(self) -> str:
...
+        :return: None
+        """
+        self._incoming_messages.extend([message])
+
 
 class Dialogues:
     """The dialogues class keeps track of all dialogues for an agent."""
 
-    def __init__(self, agent_address: Address = "") -> None:
+    def __init__(self, concrete_dialogue_class, agent_address: Address = "") -> None:
```
Comment (5A11) [inline aea/helpers/dialogue/base.py:571]: Needed this for this class's `create_dialogue` methods below.

File: aea/helpers/dialogue/base.py [hunk 6]
```diff
@@ -430,23 +613,167 @@ def update(self, message: Message) -> Optional[Dialogue]:
...
+        :param dialogue_opponent_addr: the pbk of the agent with which the dialogue is kept.
+        :param role: the agent's role
+
+        :return: the created dialogue.
+        """
+        dialogue_reference = (str(self._next_dialogue_nonce()), "")
+        dialogue_label = DialogueLabel(
+            dialogue_reference, dialogue_opponent_addr, self.agent_address
+        )
+        dialogue = self.concrete_dialogue_class.from_args(
```
Comment (5A11) [inline aea/helpers/dialogue/base.py:729]: Where the `concrete_dialogue` variable is needed.

File: aea/helpers/dialogue/base.py [hunk 7]
```diff
@@ -430,23 +613,167 @@ def update(self, message: Message) -> Optional[Dialogue]:
...
+        new_dialogue_reference = (
+            dialogue_reference[0],
+            str(self._next_dialogue_nonce()),
+        )
+        dialogue_label = DialogueLabel(
+            new_dialogue_reference, dialogue_opponent_addr, dialogue_opponent_addr
+        )
+
+        assert dialogue_label not in self.dialogues
+        dialogue = self.concrete_dialogue_class.from_args(
```
Comment (5A11) [inline aea/helpers/dialogue/base.py:762]: Where the `concrete_dialogue` variable is needed.

File: aea/helpers/dialogue/base.py [hunk 8]
```diff
@@ -154,15 +154,22 @@ def __str__(self):
             """Get the string representation."""
             return self.value
 
-    def __init__(self, dialogue_label: DialogueLabel, role: Role = None) -> None:
+    def __init__(
+        self,
+        dialogue_label: DialogueLabel,
+        agent_address: Address = None,
+        role: Role = None,
```
Comment (DavidMinarsch) [inline aea/helpers/dialogue/base.py:161]:
```suggestion
        agent_address: Optional[Address] = None,
        role: Optional[Role] = None,
```

File: aea/helpers/dialogue/base.py [hunk 9]
```diff
@@ -241,17 +267,17 @@ def last_message(self) -> Optional[Message]:
             self.last_incoming_message is not None
             and self.last_outgoing_message is not None
         ):
-            last_incoming_message_id = cast(
-                int, self.last_incoming_message.get("message_id")
-            )
-            last_outgoing_message_id = cast(
+            last_outgoing_message_message_id = cast(
                 int, self.last_outgoing_message.get("message_id")
```
Comment (DavidMinarsch) [inline aea/helpers/dialogue/base.py:271]: I would say we should add `message_id` as a property to the message base class:
```python
self._message_id = None  # type: Optional[int]
...

@property
def message_id(self) -> int:
    assert self._message_id is not None, "message_id is not set!"
    return self._message_id
```

File: aea/helpers/dialogue/base.py [hunk 10]
```diff
@@ -241,17 +267,17 @@ def last_message(self) -> Optional[Message]:
             self.last_incoming_message is not None
             and self.last_outgoing_message is not None
         ):
-            last_incoming_message_id = cast(
-                int, self.last_incoming_message.get("message_id")
-            )
-            last_outgoing_message_id = cast(
+            last_outgoing_message_message_id = cast(
```
Comment (DavidMinarsch) [inline aea/helpers/dialogue/base.py:270]: I think we should add message_id as a property to message base class..there is no harm to that:

```python
@property
def message_id(self) -> int:
    assert self.is_set("message_id"), "message_id is not set!"
    return cast(int, self.get("message_id"))
```

File: aea/helpers/dialogue/base.py [hunk 11]
```diff
@@ -241,17 +267,17 @@ def last_message(self) -> Optional[Message]:
             self.last_incoming_message is not None
             and self.last_outgoing_message is not None
         ):
```
Comment (DavidMinarsch) [inline aea/helpers/dialogue/base.py:269]: missing return type in signature

File: aea/helpers/dialogue/base.py [hunk 12]
```diff
@@ -276,77 +328,185 @@ def is_empty(self) -> bool:
...
-            is_successful = self.outgoing_safe_extend(message)
-        return is_successful
+            target = cast(int, message.get("target"))
+            last_target = self.last_message.target  # type: ignore
+            result = target == last_target + 1
+        return result
 
     @abstractmethod
-    def is_valid_next_message(self, message: "Message") -> bool:
+    def initial_performative(self):
```
Comment (DavidMinarsch) [inline aea/helpers/dialogue/base.py:459]: This means we need an abstract performative class to specify the signature correctly...

File: aea/helpers/dialogue/base.py (hunk 8 above)
Comment (5A11) [reply aea/helpers/dialogue/base.py]: Done

**Pull request merged**